docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,157 @@
+
+## Clever sharding
+Future work on sharding should aim to keep parts of the dag that are typically fetched together within the same shard.  This feature should be able to take advantage of particular chunking and layout algorithms, for example grouping together subdags representing packages when importing a file as a linux package archive.  It would also be nice to have some techniqes, possibly pluggable, available for intelligently grouping blocks of an arbitrary dag based on the dag structure.
+
+## Sharding existing dags
+This would require new features in go-ipfs if ipfs-cluster is to stream dag nodes from its local daemon.  Specifically the ipfs api would somehow need to include an endpont that provides a stream of a dag's blocks without overcommitting the daemon's resources (i.e. downloading the whole dag first if the repo is too small).  An effort that will bring sharding existing dags closer to reality is the [CAR (Certified ARchives) project](https://github.com/ipld/specs/pull/51) which aims to define a format for storing dags.  One of its goals as stated in a recent proposal is to explicitly allow users to "Traverse a large dag streamed over HTTP without downloading the entire thing."  Cluster could certainly shard dag nodes extracted from a streamed CAR file, either kept on disk or provided at a location on the network.  CAR integration with go-ipfs could potentially allow resource aware streaming of dag nodes over the ipfs daemon's api so that the sharding cluster peer need only know the dag's root hash.
+
+# Workplan
+
+There is a fair amount that needs to be built, updated and serviced to make all of this work.  Here I am focusing on basic sharding and allowing support for files that wouldn't fit in an ipfs repo.  Fault tolerance, clever sharding and support for sharding existing dags will come after this is nailed down.  
+
+* Build up a version of the `importers` (aka `dagger` aka `dex`) project that suits cluster's needs (if possible keep it general enough for general use).  In short we want a library that imports data to ipfs dags with arbitrary chunking and layout algorithms.  For our use case it is important that the interfaces support taking in a stream of data and returning a stream of blocks or dag nodes.
+* Figure out how unixFS file/directory data can be streamed (via reading from disk directly or via http POST multipart upload)
+* Implement a cluster importer component.  It will need to provide endpoints that supporting streams as input and output.
+* Create a cluster add endpoint that uses the cluster importer component to split data and create dags.  Put the output blocks into the local ipfs daemon and pin each block with cluster.  This will require figuring out the importer component's api, including the REST API part.
```

Comment:
and pins the root cid, not "each block" (i mean, not directly).

Reply:
100% with you.  I'll make it clearer (pin root recursively)